modules/tinymce/src/core/main/ts/delete/InlineFormatDelete.ts

```diff
@@ -1,28 +1,42 @@
-import { Arr, Fun, Optional } from '@ephox/katamari';
-import { SugarElement, Traverse } from '@ephox/sugar';
+import { Arr, Fun, Optional, Type } from '@ephox/katamari';
+import { PredicateExists, SugarElement, Traverse } from '@ephox/sugar';
 
 import Editor from '../api/Editor';
 import CaretPosition from '../caret/CaretPosition';
 import * as ElementType from '../dom/ElementType';
+import * as Empty from '../dom/Empty';
+import * as NodeType from '../dom/NodeType';
 import * as Parents from '../dom/Parents';
 import * as CaretFormat from '../fmt/CaretFormat';
+import { isCaretNode } from '../fmt/FormatContainer';
 import * as DeleteElement from './DeleteElement';
 import * as DeleteUtils from './DeleteUtils';
 
-const getParentInlines = (rootElm: SugarElement<Node>, startElm: SugarElement<Node>): SugarElement<Node>[] => {
+const hasOnlyOneChild = (elm: SugarElement<Node>): boolean =>
+  Traverse.childNodesCount(elm) === 1;
+
+const getParentInlines = (editor: Editor): SugarElement<Node>[] => {
+  const rootElm = SugarElement.fromDom(editor.getBody());
+  const startElm = SugarElement.fromDom(editor.selection.getStart());
   const parents = Parents.parentsAndSelf(startElm, rootElm);
   return Arr.findIndex(parents, ElementType.isBlock).fold(
     Fun.constant(parents),
     (index) => parents.slice(0, index)
-  );
+  ).filter(hasOnlyOneChild);
```

Comment:
I imagine it's something to do with inline formatting elements working that way. But any text node fragmentation, zero width cursors etc. might interfere with that. Is that a problem? Or does tinymce just normalise the issue away?

Reply:
This also raises the issue of why we are applying the one-child filter to 'self' as well (`parents = Parents.parentsAndSelf`). I guess this doesn't apply to the case when `deleteCaret` is triggered like before, but it does introduce a bug in the new `deleteRange`.

Ex:
```
<p><span style="text-decoration: underline;">a<em><strong>b</strong></em></span></p>
```
`getParentInlines` will return `[]` if cursor is at 'a', as selection start (and therefore 'self') is considered to be `<span style="text-decoration: underline;">a<em><strong>b</strong></em></span>`, which contains more than one child. This doesn't make sense as 'a' is clearly formatted with `<span style="text-decoration: underline;">`.

Reply:
Hmm. As discussed, the concern I have with this is that it doesn't make much sense to include elements that are parents of elements that have more than one children, either. Previously, the `deleteCaret` code only looked at the last element, so it wasn't important, but we're working with all of them for `deleteRange` so we probably need to exclude anything that is an ancestor of something with more than one child.

I'm also not sure if that's really accurate, either - but I don't know enough about what formatting structures we'd expect.

Reply:
Looking at the tests, I think the goal of `getParentInlines` is to find all single-child inline parents _until_ a multi-child inline parent or block. This is something the `filter` statement doesn't exactly achieve.

For example, with cursor set to `[ 0, 0, 0, 0, 0 ], 1` (between 'a' and 'b') in
```
<p><s><span style="text-decoration: underline;"><em>a</em></span>b</s></p>
```
`getParentInlines` returns 
![image](https://user-images.githubusercontent.com/56149628/205810197-66b00bc9-13ae-4bcb-be0b-ed3146cd0324.png)
Then, we take the last parent inline, the `<span ...>`, and check `DeleteUtils.willDeleteLastPositionInElement` on it. If it returns true (which it does since there is only a single character within), then the `<span ...>` gets replaced by a caret with `<span ...>` and `<em>` formats.

However, like you said Morgan, if we add an outer ancestor `<strong>` such that it has a single child,
```
<p><strong><s><span style="text-decoration: underline;"><em>a</em></span>b</s></strong></p>
```
`getParentInlines` returns
![image](https://user-images.githubusercontent.com/56149628/205811765-0b02d4eb-eace-4cf4-b8cb-8fd81459c235.png)
skipping the `<s>` which contains two children.
We check the last parent inline, the `<strong>` in `DeleteUtils.willDeleteLastPositionInElement`, which returns false as there is still a 'b' after deletion within `<strong>`, and the override is not executed. However, we ended up deleting `a` with the `<span ...>` underline and `<em>` formats, which are not retained and typing again will produce the interpreted tags like `<u>`.

This is why I propose we replace the return statement with something such as:
```
  return Arr.findIndex(parents, (x) => ElementType.isBlock(x) || !hasOnlyOneChild(x)).fold(
    Fun.constant(parents),
    (index) => parents.slice(0, index)
  );
```
so that we find all single-child inline parents _until_ a multi-child inline parent or block as required.

This is just to fix up `deleteCaret`, and my `deleteRange` will need to use something that omits `hasOnlyOneChild` completely.

Reply:
That sounds really good @mkzhx, and thanks for being so thorough with your investigation!

Reply:
Added `getParentInlinesUntilMultichildInline` (used by `deleteCaret`) to differentiate from `getParentInlines` (used by `deleteRange`).